src/test/scala/is/hail/asm4s/CodeWithRefsSuite.scala

```diff
@@ -0,0 +1,45 @@
+package is.hail.asm4s
+
+import is.hail.check.{Gen, Prop}
+import org.scalatest.testng.TestNGSuite
+import org.testng.annotations.Test
+import is.hail.asm4s.CodeM._
+import is.hail.asm4s.Code._
+
+object CodeWithRefsSuite {
+  def toISeq(a: Array[Int]): IndexedSeq[Int] = a
+}
+
+class CodeWithRefsSuite extends TestNGSuite {
+
+  @Test def sum() {
+    val m = for {
+      a <- newVar[Array[Int]](Code.newArray[Int](3))
+      _ <- a.update(0, 1)
+      _ <- a.update(1, 2)
+      _ <- a.update(2, 3)
+      i <- newVar[Int](0)
+      s <- newVar[Int](0)
+      _ <- whileLoop(i < a.length(),
+        Code(s := s + a(i), i := i + 1))
+    } yield s: Code[Int]
+
+    val f = m.run[({type F[R] = AsmFunction0[R]})#F, Int](FunctionBuilder.functionBuilder[Int])
+
+    assert(f() == 6)
+  }
+
+  @Test def sumFunction() {
+    val m = for {
+      a <- getArg[Array[Int]](0)
+      i <- newVar[Int](0)
+      s <- newVar[Int](0)
+      _ <- whileLoop(i < a.length(),
+        Code(s := s + a(i), i := i + 1))
+    } yield s: Code[Int]
```

Comment:
I'll make the CodeM[Unit] change, but I think it should actually be `Code._return` which is implemented as:

```scala
object Code {
  // ...
  def _return[T](c: Code[T])(implicit tti: TypeInfo[T]): Code[Unit] =
    Code(c, Code(new InsnNode(tti.returnOp)))
}
```

Usage will be identical to putting it on `CodeM` because of the implicit conversion.